src/comestible_inv_listitem.cpp

```diff
@@ -0,0 +1,771 @@
+                retval = d1->volume > d2->volume;
+                return d1->volume != d2->volume;
+            };
+        case COLUMN_SORTBY_CATEGORY:
+            return []( const itm * d1, const itm * d2, bool & retval ) {
+                assert( d1->cat != nullptr );
+                assert( d2->cat != nullptr );
+                if( d1->cat != d2->cat ) {
+                    retval = *d1->cat < *d2->cat;
+                    return true;
+                } else if( d1->is_category_header() ) {
+                    retval = true;
+                    return true;
+                } else if( d2->is_category_header() ) {
+                    retval = false;
+                    return true;
+                }
+                retval = false;
+                return false;
+            };
+        default:
+            return nullptr;
+    }
+}
+
+comestible_inv_listitem::comestible_inv_listitem( const std::list<item *> &list, int index,
+        comestible_inv_area *area, bool from_vehicle ) :
+    idx( index ),
+    area( area ),
+    id( list.front()->typeId() ),
+    items( list ),
+    name( list.front()->tname( list.size() ) ),
+    name_without_prefix( list.front()->tname( 1, false ) ),
+    autopickup( get_auto_pickup().has_rule( list.front() ) ),
+    stacks( list.size() ),
+    volume( list.front()->volume() * stacks ),
+    weight( list.front()->weight() * stacks ),
+    cat( &list.front()->get_category() ),
+    from_vehicle( from_vehicle )
+{
+    init( g->u.get_consumable_from( *list.front() ) );
+}
+
+void comestible_inv_listitem::init( const item &it )
+{
+    menu_color = it.color_in_inventory();
+    menu_color_dark = c_dark_gray;
+    assert( stacks >= 1 );
+}
+
+comestible_inv_listitem::comestible_inv_listitem( const item_category *cat )
+    : idx()
+    , area()
+    , id( "null" )
+    , autopickup()
+    , stacks()
+    , cat( cat )
+{
+    if( cat != nullptr ) {
+        name = cat->name();
+    }
+    menu_color_dark = c_dark_gray;
+}
+
+bool comestible_inv_listitem::is_category_header() const
+{
+    return items.empty() && cat != nullptr;
+}
+
+bool comestible_inv_listitem::is_item_entry() const
+{
+    return !items.empty();
+}
+
+
+/*******    FOOD    *******/
+
+
+comestible_inv_listitem_food::comestible_inv_listitem_food( const item_category *cat ):
+    comestible_inv_listitem( cat )
+{
+}
+
+comestible_inv_listitem_food::comestible_inv_listitem_food( const std::list<item *> &list,
+        int index,
+        comestible_inv_area *area, bool from_vehicle ): comestible_inv_listitem( list, index, area,
+                    from_vehicle )
+{
+    init( g->u.get_consumable_from( *list.front() ) );
+}
+
+void comestible_inv_listitem_food::init( const item &it )
+{
+    player &p = g->u;
+    islot_comestible edible = get_edible_comestible( p, it );
+
+    menu_color = it.color_in_inventory();
+
+    // statuses
+    const auto &comest = it.get_comestible();
```

Comment:
I assume you meant to get the `edible` instead of querying `it` directly here (`get_edible_comestible` queries it anyway).

Reply:
I don't know what the difference is between `get_edible_comestible( p, it );` and `it.get_comestible();` Edible and comestible are synonyms and it's not documented which function I should use when. 
This code is copied from different places, so not sure what it actually needs to be. The functions seem different on the inside, so I don't know the consequences if we just replace `comest` with `edible`

Reply:
> I don't know what the difference is between 

Than look at the source of their implementation.

> This code is copied from different places,

Copying code without understanding what it does (or even why it does it) is dangerous.

- `get_comestible` returns an optional value, dereferencing the optional (as you do below via `comest->`) will cause UB when it does not actually contain a value. Your code does not check this, it just dereferences in the assumption that it will always contain a value. That leads to a crash when encountering non-food items (where `comest` is empty).

`get_edible_comestible` (which looks like a straight copy from "game_inventory.cpp", which is bad on its own) checks for the optional value being empty and returns a dummy value in that case.

Btw. this has nothing to do with edible vs comestible.